keps/173-headless-service-per-group/README.md

```diff
@@ -0,0 +1,297 @@
+
+### Goals
+
+- Each group of leader-worker replicas can have a dedicated service
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Providing an API to configure service per replicas.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user running a large scale LWS job, I want to have each leader-worker group with its own service.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
+ // HeadlessServicePerReplica (if true) creates a headless service for each leader-worker group
+ // At large scale, there can be an issue where there are too many pods on a single service.
+ // This creates a headless service for each replica group.
+ // +optional
+ HeadlessServicePerReplica *bool `json:"headlessServicePerReplica,omitempty"`
+}
+```
+
+### Implementation
+
+With HeadlessServicePerReplica true, we will create a headless service per replica.
+The name of the service will be `{lws.Name}-{replica-index}`.
+Creation of the services will be in a loop of replicas.
+
+Currently, LWS will add an environment variable to all pods that points to the headless service.
```

Comment:
it is not to the headless service, it is to the dns record. Depending on how the headless service is created, the hostname will be different, and we should clarify that.